runtime/coverage.go

```diff
@@ -20,56 +20,229 @@ package runtime
 
 import (
 	"encoding/json"
+	"fmt"
+	"sort"
 
+	"github.com/onflow/cadence/runtime/ast"
 	"github.com/onflow/cadence/runtime/common"
 )
 
-// LocationCoverage records coverage information for a location
+// LocationCoverage records coverage information for a location.
 type LocationCoverage struct {
-	LineHits map[int]int `json:"line_hits"`
+	// Contains hit count for each line on a given location.
+	// A hit count of 0 means the line was not covered.
+	LineHits map[int]int
+	// Total number of statements on a given location.
+	Statements int
 }
 
+// AddLineHit increments the hit count for the given line.
 func (c *LocationCoverage) AddLineHit(line int) {
 	c.LineHits[line]++
 }
 
-func NewLocationCoverage() *LocationCoverage {
+// Percentage returns a string representation of the covered
+// statements percentage. It is defined as the ratio of covered
+// lines over the total statements for a given location.
+func (c *LocationCoverage) Percentage() string {
+	coveredLines := c.CoveredLines()
+	if coveredLines > c.Statements {
+		// We saturate the percentage at 100%, when the inspector
+		// fails to correctly count all statements for a given
+		// location.
+		coveredLines = c.Statements
```

Comment:
I was wondering how it can go beyond 100% threshold... Is it because we filter a certain set of statements when we initialize the `LocationCoverage`? i.e: here: https://github.com/onflow/cadence/blob/84b077202f093176e82931cffe2d1b2e19f0246a/runtime/coverage.go#L170-L174

and then when marking a line as covered, we don't filter them out:
https://github.com/onflow/cadence/blob/84b077202f093176e82931cffe2d1b2e19f0246a/runtime/coverage.go#L126-L130

What if we add a check to see if a given line is also excluded from the coverage (other than the location and program)? something like:
```go
func (c *LocationCoverage) AddLineHit(line int) {
	if _, included := c.LineHits[line]; included {
		c.LineHits[line]++
	}
}
```

Would that solve it?

Reply:
It seems like if we get line hits (`AddLineHit`), then those lines should have also been found / included by `InspectProgram`?

Reply:
Yes, that's my understanding. I think it should always hold that informal agreement when `AddLineHit` is invoked by the interpreter. Maybe this is just a problem for tests (when `AddLineHit` is called from outside/manually)

Reply:
My reasoning here is that the ground truth is the [newOnStatementHandler](https://github.com/onflow/cadence/blob/master/runtime/environment.go#L621-L625), which means that every call of `AddLineHit` from the callback, should be respected. The `InspectProgram` may have failed, for whatever reason, to find that specific line. But this is a good insight to solidify its implementation, and debug the inspection failure. After @turbolent suggestion's, the implementation of `InspectProgram` seems more bullet proof, than my original version, which means that the case I described above maybe does not hold anymore, and it can just manifest in tests.
How do you think we should proceed with this?

Reply:
hmm yeah, I too think that it'll never go into this condition. Thanks for the explanation.
Maybe just add a note/comment mentioning this reasoning and that ideally, it would never reach there. Then we can just leave it as is (as a fail-safe mechanism)?

Reply:
Good point :+1: 
Updated in https://github.com/onflow/cadence/pull/2296/commits/cd4f4df3fa20b841afef4cf477f28d0508244555